### PR TITLE
FIX-#3510: Correct processing of callable `skiprows` parameter of `read_csv` function

### LINKED ISSUE #3510: `read_csv` fails with some callable `skiprows` parameters
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Any
- **Modin version**: c2b399a0d69baa519dab67acd715061e93986b22
- **Python version**: 3.8.12
- **Code we can use to reproduce**:
```
import os
import modin.pandas as pd
from modin.config import NPartitions

NPartitions.put(4)

test_filename = "test.csv"
data = """col1,col2
1,2
3,4
5,6
7,8
"""

try :
    with open(test_filename, "w") as f:
        f.write(data)

    df = pd.read_csv(test_filename, skiprows=lambda x: x in [2,3])

finally:
    os.remove(test_filename)
```

### Source code / logs

```
Traceback (most recent call last):
  File "test.py", line 1189, in <module>
    df = pd.read_csv(test_filename, skiprows=lambda x: x in [2,3])
  File "/modin/modin/pandas/io.py", line 131, in read_csv
    return _read(**kwargs)
  File "/modin/modin/pandas/io.py", line 59, in _read
    pd_obj = FactoryDispatcher.read_csv(**kwargs)
  File "/modin/modin/data_management/factories/dispatcher.py", line 178, in read_csv
    return cls.__factory._read_csv(**kwargs)
  File "/modin/modin/data_management/factories/factories.py", line 206, in _read_csv
    return cls.io_cls.read_csv(**kwargs)
  File "/modin/modin/engines/base/io/file_dispatcher.py", line 68, in read
    query_compiler = cls._read(*args, **kwargs)
  File "/modin/modin/engines/base/io/text/csv_dispatcher.py", line 151, in _read
    new_query_compiler = cls._get_new_qc(
  File "/modin/modin/engines/base/io/text/csv_dispatcher.py", line 333, in _get_new_qc
    mod_index = skiprows_md(index_range)
  File "/modin/modin/engines/base/io/text/csv_dispatcher.py", line 459, in skiprows_func
    return skiprows(x + header_size)
  File "test.py", line 1189, in <lambda>
    df = pd.read_csv(test_filename, skiprows=lambda x: x in [2,3])
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

### PR DESCRIPTION
**Pull request by amyskov**
Signed-off-by: Alexander Myskov <alexander.myskov@intel.com>

## What do these changes do?
Process `skiprows` more carefully to support different types of `skiprows` callable.

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3510  
- [x] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

**Comment by amyskov** [comment]
> The handling of `skiprows` is quite complex. Would you be willing to write a document on the logic flow for handling it for the whole process (including what is already included)? I think it would be a very helpful document.

Ok, i see, handling of `skiprows` became really complex, will add docs in the separate PR.

**Comment by vnlitvinov** [comment]
@amyskov what's the state of this PR?

**Comment by amyskov** [comment]
> @amyskov what's the state of this PR?

@vnlitvinov PR is ready for review now.

File: modin/core/io/text/text_file_dispatcher.py [hunk 1]
```diff
@@ -1037,3 +1037,33 @@ def _read(cls, filepath_or_buffer: FilePathOrBuffer, **kwargs):
...
+            Callable to check whether row index should be skipped.
+
+        Returns
+        -------
+        pandas.Index
+        """
+        try:
+            # direct `skiprows_md` is more efficient than using of
+            # map method, but in some cases it can work incorrectly, e.g.
+            # when `skiprows_md` contains `in` operator
```
Comment (vnlitvinov) [inline modin/core/io/text/text_file_dispatcher.py:1060]: what is `skiprows_md`?

File: modin/core/io/text/text_file_dispatcher.py [hunk 2]
```diff
@@ -1037,3 +1037,33 @@ def _read(cls, filepath_or_buffer: FilePathOrBuffer, **kwargs):
...
+            mask = skiprows(rows_index)
+            assert is_list_like(mask)
+        except (ValueError, TypeError, AssertionError):
+            # ValueError can be raised if `skiprows` callable contain membersip operator
+            # TypeError is raised if `skiprows` callable contain bitwise operator
```
Comment (vnlitvinov) [inline modin/core/io/text/text_file_dispatcher.py:1065]:
```suggestion
            # ValueError can be raised if `skiprows` callable contains membership operator
            # TypeError is raised if `skiprows` callable contains bitwise operator
```

File: modin/pandas/test/test_io.py [hunk 3]
```diff
@@ -1132,6 +1132,30 @@ def test_read_csv_empty_frame(self):
...
+        "skiprows",
+        [
+            lambda x: x > 20,
+            lambda x: True,
+            lambda x: x in [10, 20],
+            pytest.param(
+                lambda x: x << 10,
+                marks=pytest.mark.skipif(
+                    condition="config.getoption('--simulate-cloud').lower() != 'off'",
+                    reason="The reason of tests fail in `cloud` mode is unknown for now - issue #2340",
```
Comment (vnlitvinov) [inline modin/pandas/test/test_io.py:1145]: did you test that it really fails in the cloud mode?

Comment (amyskov) [reply modin/pandas/test/test_io.py:1145]: Yes, test in this case fails with error
`RecursionError: maximum recursion depth exceeded while calling a Python object`

**Review by vnlitvinov — APPROVED** [review]
LGTM

**Review by devin-petersohn — APPROVED** [review]
LGTM

**Pull request merged**